# Видеокурс по теории вероятностей и математической статистике
# GeekBrains
### ДЗ Урока 2
### Казанцев Виталий

### Задачи 1 и 2
Контрольная работа состоит из пяти вопросов. На каждый вопрос приведено четыре варианта ответа, один из которых правильный. Случайная величина X задаёт число правильных ответов при простом угадывании. Найдите математическое ожидание данной случайной величины. Найдите дисперсию случайной величины X.

Подсказка: постройте закон распределения случайной величины X.

### Решение.
Получается что мы имеем пять независимых испытаний с вероятностью угадывания (событие $A$) равной $\frac{1}{4}$  
Тогда вероятность, что событие $A$ не наступит ни разу, т.е. $X=0$  расчитывается следующим образом:
### $P(X=0)=C_{n}^{k}p^{k}(1-p)^{n-k}=\frac{5!}{0!(5-0)!}\cdot(\frac{1}{4})^{0}\cdot(\frac{3}{4})^{5-0}=\frac{243}{1024}$

Вероятность угадать один раз

### $P(X=1)=\frac{5!}{1!\cdot4!}\cdot(\frac{1}{4})^{1}\cdot(\frac{3}{4})^4=\frac{405}{1024}$

Два раза и т.д. до пяти:

### $P(X=2)=\frac{5!}{2!\cdot3!}\cdot(\frac{1}{4})^2\cdot(\frac{3}{4})^3=\frac{135}{512}$

### $P(X=3)=\frac{5!}{3!2!}\cdot(\frac{1}{4})^{3}\cdot(\frac{3}{4})^2=\frac{45}{512}$

### $P(X=4)=\frac{5!}{4!\cdot1!}\cdot(\frac{1}{4})^{4}\cdot(\frac{3}{4})^1=\frac{15}{1024}$

### $P(X=5)=\frac{5!}{5!\cdot0!}\cdot(\frac{1}{4})^5\cdot(\frac{3}{4})^0=\frac{1}{1024}$


<table border="3">
    <tr>
        <th>$x$</th>
        <td>$0$</td>
        <td>$1$</td>
        <td>$2$</td>
        <td>$3$</td>
        <td>$4$</td>
        <td>$5$</td>
    </tr>
    <tr>
        <th>$P(X = x)$</th>
        <td>$\frac{243}{1024}$</td>
        <td>$\frac{405}{1024}$</td>
        <td>$\frac{135}{512}$</td>
        <td>$\frac{45}{512}$</td>
        <td>$\frac{15}{1024}$</td>
        <td>$\frac{1}{1024}$</td>
    </tr>
</table>

In [1]:
import numpy as np
x_values = np.arange(0,6)
x_values

array([0, 1, 2, 3, 4, 5])

In [2]:
x_probabilites = np.array([243, 405, 270, 90, 15, 1]) / 1024
x_probabilites

array([0.23730469, 0.39550781, 0.26367188, 0.08789062, 0.01464844,
       0.00097656])

In [3]:
m = (x_values * x_probabilites).sum()
print(m)

1.25


Или по другому, поскольку n=5, а p=1/4, то
### $M(X)=5\cdot 0.25=1.25$

Теперь найдем дисперсию величины $X$, которая определяется по формуле $D=M((X-M(X))^2)$  
для этого сначала найдем квадтрат отклонения величины $X$

In [4]:
y_values = x_values - m
y_values = y_values**2
y_values

array([ 1.5625,  0.0625,  0.5625,  3.0625,  7.5625, 14.0625])

Теперь вычислим дисперсию

In [5]:
d = (y_values * x_probabilites).sum()
d

0.9375

Также дисперсию можно было вычислить по формуле 
### $D(X)=np(1-p) = 1.25\cdot0.75=0.9375$

### Ответ:
Математическое ожидание равно 1.25  
Дисперсия равна 0.9375

### Задача 3
Пользователь получает в среднем 10 писем со спамом на свой почтовый ящик за сутки. Найти число N, такое, что с вероятностью 0.95 пользователь получит не более N писем со спамом за текущий день.

### Решение:
В нашем случае распределение Пуассона принимает следующий вид
### $P(X \leq N) = \displaystyle\sum_{i = 0}^{N} \dfrac{10^{i} e^{-10}}{i!}$  
Необходимо подобрать такое $N$, при котором равенство будет равно примерно 0.95

In [6]:
def poisson_proba(k: int, lambda_: float) -> float:
    """Формула Пуассона.
    """
    
    return (lambda_ ** k) * (np.exp(-lambda_)) / np.math.factorial(k)

In [7]:
lambda_ = 10

In [8]:
j = 1    # счётчик
pus = poisson_proba(k=j, lambda_=lambda_)    # веороятность при значении X=1
while pus < 0.95:
    # На каждой итерации сравниваем значение вероятности с искомым
    j += 1
    pus = sum(poisson_proba(k=i, lambda_=lambda_) for i in range(j))
    
print(j-1)

15


In [9]:
sum(poisson_proba(k=i, lambda_=lambda_) for i in range(16))

0.9512595966960214

### Ответ:
С вероятностью в 95% пользователь получит не более 15 писем со спамом в день

### Задача 4
Производятся выстрелы по мишени. Вероятность попадания в мишень при одном выстреле равна 0.01. Сколько выстрелов нужно сделать чтобы быть уверенным с вероятностью 0.9, что хотя бы 6 раз будет совершено попадание?

Подсказка: 1) "Вероятность попасть k раз при n выстрелах" - на какое распределение это похоже? 2) А если нам нужна вероятность P(X >= k), а не P(X = k)? 3) Здесь предстоит немножко покодить.

Это похоже на биномиальное расппределение $P(X = k) = C_n^k p^k (1 - p)^{n - k}$, при $k=6$ и $p=0.01$ получаем 
### $\begin{cases}P(X=6)=0.9,\\P(X=6)=C_{n}^6\cdot0.01^{6}\cdot(1-0.01)^{n-6}\end{cases}$ $\Rightarrow\frac{n!}{6!(n-6)!}\cdot10^{-12}\cdot0.99^{n-6}=0.9, n\geq6$  
Необходимо подобрать значение $n$

In [10]:
def probability_binominal(n, k, p):
    c = np.math.factorial(n) / (np.math.factorial(k) * np.math.factorial(n - k))
    p = p**k
    probability = c * p * ((1 - p)**(n-k))
    return probability

In [11]:
n = 6
bin_p = probability_binominal(n=n, k=6, p=0.01)
while bin_p < 0.9:
    n += 1
    bin_p = probability_binominal(n=n, k=6, p=0.01)
print(n)

297


In [12]:
probability_binominal(n=297, k=6, p=0.01)

0.9060151962843556

### Ответ
Чтобы при заданных условиях попасть в мишень миниму 6 раз, необходимо произвести около 297 выстрелов